lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -306,32 +316,42 @@ streamBlocks logTrace smM logic batchSize streamWindow enqueue nodeId tipHeader
+    -- thread is killed, and it also awaits the sub-thread so will be
+    -- killed if the sub thread dies.
     requestBlocks :: IO (Async.Async (Maybe t))
-    requestBlocks = do
+    requestBlocks = mask $ \restore -> do
         convMap <- enqueue
-          (MsgRequestBlocks (S.singleton nodeId))
-          (\_ _ ->  (Conversation $ streamBlocksConversation) :|
-                   [(Conversation $ batchConversation)]
-          )
-        -- Outbound queue guarantees that the map is either size 0 or 1, since
-        -- 'S.singleton nodeId' was given to the enqueue.
-        case M.lookup nodeId convMap of
-            Just tvar -> atomically $ do
-                pStatus <- Conc.readTVar tvar
-                case pStatus of
-                    OQ.PacketEnqueued -> Conc.retry
-                    -- Somebody else arborted our call; nothing to do but
-                    -- throw.
-                    OQ.PacketAborted  -> Conc.throwSTM $ DialogUnexpected $ "streamBlocks: aborted"
-                    OQ.PacketDequeued streamThread -> pure streamThread
-            -- FIXME shouldn't have to deal with this.
-            -- One possible solution: do the block request in response to an
-            -- unsolicited header, so that's it's all done in one conversation,
-            -- and so there's no need to even track the 'nodeId'.
-            Nothing   -> throwIO $ DialogUnexpected $ "streamBlocks: did not contact given peer"
+            (MsgRequestBlocks (S.singleton nodeId))
+            (\_ _ ->  (Conversation $ streamBlocksConversation) :|
+                     [(Conversation $ batchConversation)]
+            )
+        let waitForDequeue = case M.lookup nodeId convMap of
+                Just tvar -> atomically $ do
+                    pStatus <- Conc.readTVar tvar
+                    case pStatus of
+                        OQ.PacketEnqueued -> Conc.retry
+                        OQ.PacketAborted -> Conc.throwSTM $ DialogUnexpected $ "streamBlocks: aborted"
+                        OQ.PacketDequeued streamThread -> pure streamThread
+                Nothing -> throwIO $ DialogUnexpected $ "streamBlocks: did not contact given peer"
+            -- Abort the conversation if it's not yet enqueued, or cancel
+            -- it if is enqueued.
+            abortOrCancel = case M.lookup nodeId convMap of
+                Just tvar -> join $ atomically $ do
+                    pStatus <- Conc.readTVar tvar
+                    case pStatus of
+                        OQ.PacketEnqueued -> do
+                            Conc.writeTVar tvar OQ.PacketAborted
+                            pure (pure ())
+                        OQ.PacketAborted  ->
+                            pure (pure ())
+                        OQ.PacketDequeued streamThread ->
+                            pure (Async.cancel streamThread)
+                Nothing -> pure ()
+        restore waitForDequeue `onException` abortOrCancel
```

Comment:
`Control.Exception.onException` has type `IO a -> IO b -> IO a` which means it will ignore `b`.  `abortOrCancel` return type is `IO (IO ())`, so the `Async.cancel streamThread` will be ignored; I guess the `onException` here comes from somewhere else.

Reply:
> `abortOrCancel` return type is `IO (IO ())`

No it's `IO ()`. There's a `join` in the `Just` case.

Reply:
@avieth forget about it, there's a `join` on top, which I missed.